File: api/utils/timer/timer.go
package timer 

import "time"

// VideoTimer tracks the timing of the current playing video
type VideoTimer struct {
	Start time.Time
	Progress time.Duration
	stop bool
}

// Pause pauses the timer for the currently playing video
func (t *VideoTimer) Pause() *VideoTimer {
	t.Progress = time.Since(t.Start) + t.Progress
	t.Start = time.Now()
	t.stop = true
	return t
}

// Play starts the timer for the currently playing video
func (t *VideoTimer) Play() *VideoTimer {
	t.Start = time.Now()
	t.stop = false
	return t
}

// Elapsed returns time elapsed since video start in milliseconds
func (t *VideoTimer) Elapsed() int64 {

	if t.stop {
		return t.Progress.Milliseconds()
	} else {
		elapsed := t.Progress + time.Since(t.Start)
		return elapsed.Milliseconds()
	}
}

// SeekTo sets time elapsed to the provided millisecond
func (t *VideoTimer) SeekTo(ms int64) *VideoTimer {
	t.Progress = time.Duration(ms) * time.Millisecond
	t.Start = time.Now()
	return t
}

File: client/src/features/timer/Timer.js
export default class VideoTimer {
    constructor(start, progress) {
        this.start = start;

        // Time elapsed in ms
        this.progress = progress;
        this.stop = true;
        this.interval = null;
    }

    pause() {
        let now = new Date();
        this.progress = now.getTime() - this.start.getTime() + this.progress;
        this.start = now;
        this.stop = true;
        
        // clearInterval(this.interval)
        // this.interval = null;
        return this
    }

    play() {
        this.start = new Date();
        this.stop = false;
        // if (this.interval == null) {
        //     this.interval = setInterval(() => {console.log(self.elapsed())}, 1000)
        // }
        return this
    }

    // Returns time elapsed since video started in ms
    elapsed() {
        let now = new Date();

        if (this.stop) {
            return this.progress
        } else {
            let elapsed = this.progress + (now.getTime() - this.start.getTime());
            return elapsed;
        }
    }

    seekTo(ms) {
        this.progress = ms;
        // console.log(`Time seeked to ${this.progress / 1000} seconds`);
        this.start = new Date();
        return this;
    }
}